src/Java.Interop.Tools.JavaTypeSystem/JavaModels/JavaClassModel.cs

```diff
@@ -0,0 +1,118 @@
+using System.Linq;
+
+namespace Java.Interop.Tools.JavaTypeSystem.Models
+{
+	public class JavaClassModel : JavaTypeModel
+	{
+		IDictionary<JavaTypeReference, JavaTypeReference>? generic_inheritance_mapping;
+
+		public string BaseType { get; }
+		public string BaseTypeGeneric { get; }
+		public string BaseTypeJni { get; }
+
+		public JavaTypeReference? BaseTypeReference { get; private set; }
+		public List<JavaConstructorModel> Constructors { get; } = new List<JavaConstructorModel> ();
+
+		public JavaClassModel (JavaPackage javaPackage, string javaNestedName, string javaVisibility, bool javaAbstract, bool javaFinal, string javaBaseType, string javaBaseTypeGeneric, string javaDeprecated, bool javaStatic, string jniSignature, string baseTypeJni) :
+			base (javaPackage, javaNestedName, javaVisibility, javaAbstract, javaFinal, javaDeprecated, javaStatic, jniSignature)
+		{
+			BaseType = javaBaseType;
+			BaseTypeGeneric = javaBaseTypeGeneric;
+			BaseTypeJni = baseTypeJni;
+		}
+
+		public override void Resolve (JavaTypeCollection types, List<JavaUnresolvableModel> unresolvables)
+		{
+			var type_parameters = GetApplicableTypeParameters ().ToArray ();
+
+			// Resolve base class
+			if (FullName != "java.lang.Object" && FullName != "java.lang.Throwable") {
+				try {
+					BaseTypeReference = types.ResolveTypeReference (TypeResolutionOptions.ResolveGenerics ? BaseTypeGeneric : BaseType, type_parameters);
+				} catch (JavaTypeResolutionException) {
+					unresolvables.Add (new JavaUnresolvableModel (this, BaseTypeGeneric, UnresolvableType.BaseType));
+
+					throw;
+				}
+
+				// We don't resolve reference-only types by default, so if our base class
+				// is a reference only type, we need to force it to resolve here. This will be
+				// needed later when we attempt to resolve base methods.
+				try {
+					if (BaseTypeReference.ReferencedType is JavaClassModel klass && klass.FullName != "java.lang.Object" && klass.BaseTypeReference is null && klass.IsReferencedOnly)
+						klass.Resolve (types, unresolvables);
+				} catch (JavaTypeResolutionException) {
+					// Ignore
+				}
+			}
+
+			// Resolve constructors
+			foreach (var ctor in Constructors)
+				ctor.Resolve (types, unresolvables);
+
+			base.Resolve (types, unresolvables);
+		}
+
+		public virtual void ResolveBaseMembers ()
+		{
+			var klass = BaseTypeReference?.ReferencedType as JavaClassModel;
+
+			foreach (var method in Methods)
+				method.FindBaseMethod (klass);
+
+			foreach (var nested in NestedTypes.OfType<JavaClassModel> ())
+				nested.ResolveBaseMembers ();
+		}
+
+		public IDictionary<JavaTypeReference, JavaTypeReference>? GenericInheritanceMapping {
+			get {
+				PrepareGenericInheritanceMapping ();
+				return generic_inheritance_mapping;
+			}
+		}
+
+		void PrepareGenericInheritanceMapping ()
+		{
+			if (generic_inheritance_mapping != null)
+				return; // already done.
+
+			var bt = BaseTypeReference?.ReferencedType as JavaClassModel;
```

Comment:
I'm not sure how to verify this method, as I'm not sure what `BaseTypeReference.ReferencedType` is supposed to represent.